master/internal/stream/models.go

```diff
@@ -0,0 +1,233 @@
+package stream
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/uptrace/bun"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/stream"
+)
+
+const (
+	// ModelsDeleteKey specifies the key for delete models.
+	ModelsDeleteKey = "models_deleted"
+	// ModelsUpsertKey specifies the key for upsert models.
+	ModelsUpsertKey = "model"
+	// modelChannel specifies the channel to listen to model events.
+	modelChannel = "stream_model_chan"
+)
+
+// ModelMsg is a stream.Msg.
+//
+// determined:stream-gen source=server delete_msg=ModelsDeleted
+type ModelMsg struct {
+	bun.BaseModel `bun:"table:models"`
+
+	// immutable attributes
+	ID int `bun:"id,pk" json:"id"`
+
+	// mutable attributes
+	Name            string    `bun:"name" json:"name"`
+	Description     string    `bun:"description" json:"description"`
+	Archived        bool      `bun:"archived" json:"archived"`
+	CreationTime    time.Time `bun:"creation_time" json:"creation_time"`
+	Notes           JSONB     `bun:"notes" json:"notes"`
+	WorkspaceID     int       `bun:"workspace_id" json:"workspace_id"`
+	UserID          int       `bun:"user_id" json:"user_id"`
+	LastUpdatedTime time.Time `bun:"last_updated_time" json:"last_updated_time"`
+	Metadata        JSONB     `bun:"metadata" json:"metadata"`
+	Labels          JSONB     `bun:"labels" json:"labels"`
+
+	// metadata
+	Seq int64 `bun:"seq" json:"seq"`
+}
+
+// SeqNum gets the SeqNum from a ModelMsg.
+func (pm *ModelMsg) SeqNum() int64 {
+	return pm.Seq
+}
+
+// UpsertMsg creates a model stream upsert message.
+func (pm *ModelMsg) UpsertMsg() stream.UpsertMsg {
+	return stream.UpsertMsg{
+		JSONKey: ModelsUpsertKey,
+		Msg:     pm,
+	}
+}
+
+// DeleteMsg creates a model stream delete message.
+func (pm *ModelMsg) DeleteMsg() stream.DeleteMsg {
+	deleted := strconv.FormatInt(int64(pm.ID), 10)
+	return stream.DeleteMsg{
+		Key:     ModelsDeleteKey,
+		Deleted: deleted,
+	}
+}
+
+// ModelSubscriptionSpec is what a user submits to define a Model subscription.
+//
+// determined:stream-gen source=client
+type ModelSubscriptionSpec struct {
```

Comment:
I don't see an issue with allow subscription based on user id; assuming RBAC is implemented in EE. 

I would be hesitant to allow label-based subscriptions for the same reason as above, especially since there doesn't seem to be a limit on the length of a tag.